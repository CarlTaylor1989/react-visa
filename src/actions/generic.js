import * as types from './genericTypes';

export function changeScreen(screen) {
  return {
    type: types.CHANGE_SCREEN,
    screen
  };
}

export function resetScreen() {
  return {
    type: types.RESET_SCREEN
  };
}

export function suspendDataParsed(dataReady) {
  return {
    type: types.SUSPEND_DATA_PARSED,
    dataReady
  };
}

export function setGenericData(data) {
  return {
    type: types.SET_GENERIC_DATA,
    data
  };
}

export function updateLaunched() {
  return {
    type: types.UPDATE_LAUNCHED
  };
}

export function setScreenReferrer(screenReferrer) {
  return {
    type: types.SET_SCREEN_REFERRER,
    screenReferrer
  };
}

export function setPopupReferrer(popupReferrer) {
  return {
    type: types.SET_POPUP_REFERRER,
    popupReferrer
  };
}

export function setProductPageInactive(productPageInactive) {
  return {
    type: types.SET_PRODUCT_PAGE_INACTIVE,
    productPageInactive
  };
}

export function toggleSettingsBtnState() {
  return {
    type: types.TOGGLE_SETTINGS_BTN_STATE
  };
}

export function disableFeedback() {
  return {
    type: types.DISABLE_FEEDBACK
  };
}

export const setClientIntroSeen = () => ({
  type: types.SET_CLIENT_INTRO_SEEN
});

export const setNetworkIntroSeen = () => ({
  type: types.SET_NETWORK_INTRO_SEEN
});

export const setIntroSeqCompleted = () => ({
  type: types.SET_INTRO_SEQ_COMPLETED
});
