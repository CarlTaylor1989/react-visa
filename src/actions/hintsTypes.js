export const SET_CURRENT_HINT_GROUP = 'SET_CURRENT_HINT_GROUP';
export const SET_HINT_GROUP_INDEX = 'SET_HINT_GROUP_INDEX';
export const SET_HINTS_PAUSED = 'SET_HINTS_PAUSED';
