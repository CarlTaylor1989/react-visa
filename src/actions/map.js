import * as types from './mapTypes';

export function setChallenges(challenges) {
  return {
    type: types.SET_CHALLENGES,
    challenges
  };
}

export function setFamilies(families) {
  return {
    type: types.SET_FAMILIES,
    families
  };
}

export function initialiseMapState(initialised) {
  return {
    type: types.MAP_STATE_INITIALISED,
    initialised
  };
}

export function updateChallengeStatus(id, familyId, status) {
  return {
    type: types.UPDATE_CHALLENGE_STATUS,
    id,
    familyId,
    status
  };
}

export function updateFamilyStatus(id, status) {
  return {
    type: types.UPDATE_FAMILY_STATUS,
    id,
    status
  };
}

export function setLastChallengesModified(lastChallengesModified) {
  return {
    type: types.SET_LAST_CHALLENGES_MODIFIED,
    lastChallengesModified
  };
}

export function setLastProductCompleted(lastProductCompleted) {
  return {
    type: types.SET_LAST_PRODUCT_COMPLETED,
    lastProductCompleted
  };
}

export function resetLastProductCompleted() {
  return {
    type: types.RESET_LAST_PRODUCT_COMPLETED
  };
}

export function setCurrentFamilyProduct(currentMap, currentFamily, currentProduct) {
  return {
    type: types.SET_CURRENT_FAMILY_PRODUCT,
    currentMap,
    currentFamily,
    currentProduct
  };
}

export function addPerfectChallenge(challenge) {
  return {
    type: types.ADD_PERFECT_CHALLENGE,
    challenge
  };
}

export function setCompletedProducts(products) {
  return {
    type: types.SET_COMPLETED_PRODUCTS,
    products
  };
}

export function addCompletedProduct(productId) {
  return {
    type: types.ADD_COMPLETED_PRODUCT,
    productId
  };
}

export function setCompletedChallenges(challenges) {
  return {
    type: types.SET_COMPLETED_CHALLENGES,
    challenges
  };
}

export function addCompletedChallenge(challengeId) {
  return {
    type: types.ADD_COMPLETED_CHALLENGE,
    challengeId
  };
}

export function resetProgress() {
  return {
    type: types.RESET_PROGRESS
  };
}

export function setProductPageViewedSession() {
  return {
    type: types.SET_PRODUCT_PAGE_VIEWED_SESSION
  };
}
