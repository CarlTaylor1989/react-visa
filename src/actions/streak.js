import * as types from './streakTypes';

export function increaseStreak() {
  return {
    type: types.INCREASE_STREAK
  };
}

export function setStreak(data) {
  return {
    type: types.SET_STREAK,
    data
  };
}

export function resetStreakProgress(challengeId, familyId) {
  return {
    type: types.RESET_STREAK_PROGRESS,
    challengeId,
    familyId
  };
}

export function checkStreakStatus() {
  return {
    type: types.CHECK_STREAK_STATUS
  };
}
