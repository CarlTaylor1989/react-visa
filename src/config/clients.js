export const REQUIREMENT_MAX_ATTEMPTS = 3;
export const CLIENT_LOCKED = { // in seconds
  1: 60, // 1 minute
  2: 120, // 2 minutes
  3: 300, // 5 minutes
  4: 600, // 10 minutes
  max: 600, // 10 minutes
};
export const STEP1_SCORES = {
  1: 50000,
  2: 100000,
  3: 150000
};
export const STEP2_SCORES = {
  1: 100000,
  2: 200000,
  3: 300000
};

export default {
  client1: {
    difficulty: 1,
    statements: ['st1', 'st2', 'st5', 'st9'],
    solutions: ['prod1', 'prod2'],
    challenges: 1,
    goals: 1,
    opportunities: 2
  },
  client2: {
    difficulty: 1,
    statements: ['st2', 'st4', 'st6', 'st9', 'st11'],
    solutions: ['prod1', 'prod2'],
    challenges: 2,
    goals: 3,
    opportunities: 0
  },
  client3: {
    difficulty: 3,
    statements: ['st1', 'st2', 'st5', 'st9'],
    solutions: ['prod1', 'prod2'],
    challenges: 1,
    goals: 1,
    opportunities: 2
  },
  client4: {
    difficulty: 1,
    statements: ['st2', 'st3', 'st6', 'st9', 'st11', 'st12'],
    solutions: ['prod1', 'prod2'],
    challenges: 2,
    goals: 4,
    opportunities: 0
  },
  client5: {
    difficulty: 2,
    statements: ['st1', 'st2', 'st5', 'st9'],
    solutions: ['prod1', 'prod2'],
    challenges: 1,
    goals: 1,
    opportunities: 2
  },
  client6: {
    difficulty: 3,
    statements: ['st1', 'st2', 'st5', 'st9'],
    solutions: ['prod1', 'prod2'],
    challenges: 1,
    goals: 1,
    opportunities: 2
  },
  client7: {
    difficulty: 1,
    statements: ['st1', 'st2', 'st5', 'st9'],
    solutions: ['prod1', 'prod2'],
    challenges: 1,
    goals: 1,
    opportunities: 2
  },
  client8: {
    difficulty: 2,
    statements: ['st1', 'st2', 'st5', 'st9'],
    solutions: ['prod1', 'prod2'],
    challenges: 1,
    goals: 1,
    opportunities: 2
  },
  client9: {
    difficulty: 3,
    statements: ['st1', 'st2', 'st5', 'st9'],
    solutions: ['prod1', 'prod2'],
    challenges: 1,
    goals: 1,
    opportunities: 2
  },
  client10: {
    difficulty: 1,
    statements: ['st1', 'st2', 'st5', 'st9'],
    solutions: ['prod1', 'prod2'],
    challenges: 1,
    goals: 1,
    opportunities: 2
  }
};
