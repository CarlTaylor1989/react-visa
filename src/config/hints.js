export const HINT_DIAGNOSTIC = 'diagnostic';
export const HINT_NETWORK = 'network';
export const HINT_PRODUCT = 'product';
export const HINT_CLIENT_SELECTION = 'clientselection';
export const HINT_CLIENT_PROPOSAL = 'clientproposal';

export const HINT_TIME = 20000;

export default [
  {
    id: 'h1',
    group: HINT_DIAGNOSTIC
  },
  {
    id: 'h2',
    group: HINT_NETWORK
  },
  {
    id: 'h3',
    group: HINT_NETWORK
  },
  {
    id: 'h4',
    group: HINT_NETWORK
  },
  {
    id: 'h5',
    group: HINT_NETWORK
  },
  {
    id: 'h6',
    group: HINT_NETWORK
  },
  {
    id: 'h7',
    group: HINT_NETWORK
  },
  {
    id: 'h8',
    group: HINT_NETWORK
  },
  {
    id: 'h9',
    group: HINT_PRODUCT
  },
  {
    id: 'h10',
    group: HINT_PRODUCT
  },
  {
    id: 'h11',
    group: HINT_CLIENT_SELECTION
  },
  {
    id: 'h12',
    group: HINT_CLIENT_PROPOSAL
  },
  {
    id: 'h13',
    group: HINT_CLIENT_PROPOSAL
  }
];
