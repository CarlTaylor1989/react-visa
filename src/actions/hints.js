import * as types from './hintsTypes';

export function setCurrentHintGroup(group) {
  return {
    type: types.SET_CURRENT_HINT_GROUP,
    group
  };
}

export function setHintGroupIndex(hints) {
  return {
    type: types.SET_HINT_GROUP_INDEX,
    hints
  };
}

export function setHintsPaused(paused) {
  return {
    type: types.SET_HINTS_PAUSED,
    paused
  };
}
