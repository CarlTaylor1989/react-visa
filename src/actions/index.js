import { push } from 'connected-react-router';
import * as AchievementActions from './achievements';
import * as AudioActions from './audio';
import * as ClientActions from './client';
import * as DiagnosticActions from './diagnostic';
import * as GenericActions from './generic';
import * as HintsActions from './hints';
import * as MapActions from './map';
import * as PromptActions from './prompts';
import * as ScoreActions from './score';
import * as ScormActions from './scorm';
import * as SettingsActions from './settings';
import * as StreakActions from './streak';
import * as TutorialActions from './tutorial';

const ActionCreators = Object.assign(
  {},
  AchievementActions,
  AudioActions,
  ClientActions,
  DiagnosticActions,
  GenericActions,
  HintsActions,
  MapActions,
  PromptActions,
  ScoreActions,
  ScormActions,
  SettingsActions,
  StreakActions,
  TutorialActions,
  { push }
);

export default ActionCreators;
