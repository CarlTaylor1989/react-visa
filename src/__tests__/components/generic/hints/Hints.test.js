/* global toJson */
import React from 'react'; // eslint-disable-line no-unused-vars
import { cloneDeep } from 'lodash';
import {
  Hints,
  mapStateToProps,
  mapDispatchToProps
} from '../../../../components/generic/hints/Hints';
import * as actions from '../../../../actions/hints';
import * as types from '../../../../actions/hintsTypes';
import { initialState as hintsInitialState } from '../../../../reducers/hints';
import { initialState as mapInitialState } from '../../../../reducers/map';
import { initialState as streakInitialState } from '../../../../reducers/streak';
import { initialState as genericInitialState } from '../../../../reducers/generic';
import { initialState as clientProposalIinitialState } from '../../../../reducers/clients';
import {
  getHintId,
  getHintIndexFromState,
  getNextHintIndex,
  hintAccessible
} from '../../../../lib/hints';

jest.useFakeTimers();

jest.mock('../../../../lib/hints');

let mockProps;
beforeEach(() => {
  jest.resetAllMocks();
  mockProps = {
    bonusStreak: streakInitialState.progress,
    clients: clientProposalIinitialState.clients,
    clientShowing: clientProposalIinitialState.clientShowing,
    displayReqFeedback: clientProposalIinitialState.displayReqFeedback,
    currentGroup: hintsInitialState.currentGroup,
    completedChallenges: mapInitialState.completedChallenges,
    completedProducts: mapInitialState.completedProducts,
    hints: hintsInitialState.hints,
    paused: hintsInitialState.paused,
    productPageInactive: false,
    setHintGroupIndex: jest.fn()
  };
});

test('renders Hints component', () => {
  const wrapper = shallow(<Hints {...mockProps} />);
  expect(toJson(wrapper, { noKey: false })).toMatchSnapshot();
});

describe('method componentDidMount()', () => {
  test('setHint() called', () => {
    // setup
    const wrapper = shallow(<Hints {...mockProps} />, {
      disableLifecycleMethods: true,
    });
    wrapper.instance().setHint = jest.fn();

    // function under test
    wrapper.instance().componentDidMount();

    // expectation
    expect(wrapper.instance().setHint).toHaveBeenCalled();
  });

  test('timer is called', () => {
    // setup
    const wrapper = shallow(<Hints {...mockProps} />, {
      disableLifecycleMethods: true,
    });
    wrapper.instance().setHint = jest.fn();
    wrapper.instance().componentDidMount();
    jest.runOnlyPendingTimers();

    // expectation
    expect(wrapper.instance().setHint).toHaveBeenCalled();
  });
});

describe('method componentDidUpdate()', () => {
  test('method setHint() is not called', () => {
    // setup
    mockProps.paused = true;
    const wrapper = shallow(<Hints {...mockProps} />, {
      disableLifecycleMethods: true,
    });
    wrapper.instance().setHint = jest.fn();

    // function under test
    wrapper.instance().componentDidUpdate(mockProps);

    // expectation
    expect(wrapper.instance().setHint).not.toHaveBeenCalled();
  });

  test('method setHint() is called after unpaused or different hint group', () => {
    // setup
    const previousProps = cloneDeep(mockProps);
    previousProps.paused = true;
    mockProps.paused = false;
    const wrapper = shallow(<Hints {...mockProps} />, {
      disableLifecycleMethods: true,
    });
    wrapper.instance().setHint = jest.fn();

    // function under test
    wrapper.instance().componentDidUpdate(previousProps);

    // expectation
    expect(wrapper.instance().setHint).toHaveBeenCalled();
  });

  test(`method setNextHint() is called after the hint group
  is changed from an existing one`, () => {
    // setup
    const previousProps = cloneDeep(mockProps);
    previousProps.currentGroup = 'diagnostic';
    mockProps.paused = false;
    mockProps.currentGroup = 'proposal';
    const wrapper = shallow(<Hints {...mockProps} />, {
      disableLifecycleMethods: true,
    });
    wrapper.instance().setNextHint = jest.fn();

    // function under test
    wrapper.instance().componentDidUpdate(previousProps);

    // expectation
    expect(wrapper.instance().setNextHint).toHaveBeenCalled();
  });

  test('method setNextHint() is called when completedChallenge is changed', () => {
    // setup
    mockProps.currentGroup = 'network';
    const previousProps = cloneDeep(mockProps);
    mockProps.completedChallenges = ['one'];
    getHintId.mockImplementation(() => 'h2');
    getHintIndexFromState.mockImplementation(() => 0);
    const wrapper = shallow(<Hints {...mockProps} />, {
      disableLifecycleMethods: true,
    });
    wrapper.instance().setNextHint = jest.fn();

    // function under test
    wrapper.instance().componentDidUpdate(previousProps);

    // expectation
    expect(wrapper.instance().setNextHint).toHaveBeenCalled();
  });

  test('method setNextHint() is called when completedProducts is changed', () => {
    // setup
    mockProps.currentGroup = 'network';
    const previousProps = cloneDeep(mockProps);
    mockProps.completedProducts = ['one'];
    getHintId.mockImplementation(() => 'h2');
    getHintIndexFromState.mockImplementation(() => 0);
    const wrapper = shallow(<Hints {...mockProps} />, {
      disableLifecycleMethods: true,
    });
    wrapper.instance().setNextHint = jest.fn();

    // function under test
    wrapper.instance().componentDidUpdate(previousProps);

    // expectation
    expect(wrapper.instance().setNextHint).toHaveBeenCalled();
  });
});

test('clearInterval() called from method componentWillUnmount()', () => {
  const wrapper = shallow(<Hints {...mockProps} />, {
    disableLifecycleMethods: true,
  });
  wrapper.instance().componentWillUnmount();
  expect(clearInterval).toHaveBeenCalled();
});

describe('function setHint()', () => {
  test('hintAccessible() returns true for accessible hint', () => {
    // setup
    const proposal = 'h12';
    mockProps.currentGroup = 'proposal';
    getHintIndexFromState.mockImplementation(() => 0);
    getHintId.mockImplementation(() => proposal);
    hintAccessible.mockImplementation(() => true);
    const wrapper = shallow(<Hints {...mockProps} />);

    // function under test
    wrapper.instance().setHint();

    // expectation
    expect(wrapper.instance().state.hintId).toEqual(proposal);
  });

  test('hintAccessible() returns false for accessible hint', () => {
    // setup
    const proposal = '';
    mockProps.currentGroup = 'proposal';
    getHintIndexFromState.mockImplementation(() => 0);
    getHintId.mockImplementation(() => null);
    hintAccessible.mockImplementation(() => false);
    const wrapper = shallow(<Hints {...mockProps} />);

    // function under test
    wrapper.instance().setHint();

    // expectation
    expect(wrapper.instance().state.hintId).toEqual(proposal);
  });
});

describe('method setNextHint()', () => {
  test('function getHintIndexFromState() called for diagnostic', () => {
    // setup
    mockProps.currentGroup = 'diagnostic';
    const wrapper = shallow(<Hints {...mockProps} />);

    // function under test
    wrapper.instance().setNextHint();

    // expectation
    expect(getHintIndexFromState).toHaveBeenCalledWith(
      mockProps.hints,
      mockProps.currentGroup
    );
  });

  test('function getNextHintIndex() called', () => {
    // setup
    mockProps.currentGroup = 'proposal';
    const wrapper = shallow(<Hints {...mockProps} />);
    getHintIndexFromState.mockImplementation(() => 1);

    // function under test
    wrapper.instance().setNextHint();

    // expectation
    expect(getNextHintIndex).toHaveBeenCalled();
  });

  test('function getNextHintIndex() not called', () => {
    // setup
    mockProps.currentGroup = 'proposal';
    const wrapper = shallow(<Hints {...mockProps} />);
    getHintIndexFromState.mockImplementation(() => null);

    // function under test
    wrapper.instance().setNextHint();

    // expectation
    expect(getNextHintIndex).not.toHaveBeenCalled();
  });

  test('function setHintGroupIndex() called', () => {
    // setup
    mockProps.currentGroup = 'proposal';
    const wrapper = shallow(<Hints {...mockProps} />);
    getHintIndexFromState.mockImplementation(() => 2);
    getNextHintIndex.mockImplementation(() => 3);

    // function under test
    wrapper.instance().setNextHint();

    // expectation
    expect(mockProps.setHintGroupIndex).toHaveBeenCalled();
  });
});

test('should mapStateToProps', () => {
  const state = {
    clientProposalData: clientProposalIinitialState,
    hintsData: hintsInitialState,
    mapData: mapInitialState,
    streakData: streakInitialState,
    genericData: genericInitialState
  };
  const props = mapStateToProps(state);
  expect(props).toEqual({
    currentGroup: state.hintsData.currentGroup,
    paused: state.hintsData.paused,
    hints: state.hintsData.hints,
    clients: state.clientProposalData.clients,
    clientShowing: state.clientProposalData.clientShowing,
    displayReqFeedback: state.clientProposalData.displayReqFeedback,
    completedChallenges: state.mapData.completedChallenges,
    completedProducts: state.mapData.completedProducts,
    bonusStreak: state.streakData.progress,
    productPageInactive: state.genericData.productPageInactive
  });
});

test('should mapDispatchToProps', () => {
  // setup
  const dispatch = jest.fn();
  const hints = { group: 'proposal', index: 1 };
  jest.spyOn(actions, 'setHintGroupIndex').mockReturnValue('setHintGroupIndex()');
  const props = mapDispatchToProps(dispatch);

  // function under test
  props.setHintGroupIndex(hints);

  // expectation
  expect(dispatch).toHaveBeenCalledWith({
    type: types.SET_HINT_GROUP_INDEX,
    hints
  });
});
