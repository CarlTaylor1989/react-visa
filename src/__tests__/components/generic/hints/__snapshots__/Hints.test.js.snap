// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders Hints component 1`] = `
<div
  className="hints"
>
  <div
    className="hint"
  >
    <span
      className="icon"
    />
    <Translate
      tag="span"
      value="generic.hint"
    />
  </div>
  <div
    className="hintText"
  >
    <span
      className="icon"
    />
  </div>
</div>
`;
